scopesim/optics/optical_train.py

```diff
@@ -185,6 +195,83 @@ def observe(self, orig_source, update=True, **kwargs):
 
         self._last_source = source
 
+
+    def prepare_source(self, source):
+        """
+        Prepare source for observation
+
+        The method is currently applied to cube fields only.
+        The source data are converted to internally used units (PHOTLAM).
+        The source data are interpolated to the waveset used by the FieldOfView
+        This is necessary when the source data are sampled on a coarser grid
+        than used internally, or if the source data are sampled on irregular
+        wavelengths.
+        For cube fields, the method assumes that the wavelengths at which the
+        cube is sampled is provided explicitely as attribute `wave` if the cube
+        ImageHDU.
+        """
+        # Convert to PHOTLAM per arcsec2
+        # ..todo: this is not sufficiently general
+
+        for cube in source.cube_fields:
+            header, data, wave = cube.header, cube.data, cube.wave
+
+            # Need to check whether BUNIT is per arcsec2 or per pixel
+            inunit = u.Unit(header['BUNIT'])
+            data = data.astype(np.float32) * inunit
+            factor = 1
+            for un, power in zip(inunit.bases, inunit.powers):
+                if (un**power).is_equivalent(u.arcsec**(-2)):
+                    conversion =(un**power).to(u.arcsec**(-2)) / un**power
+                    data *= conversion
+                    factor = u.arcsec**(-2)
+
+            data = data.to(PHOTLAM,
+                           equivalencies=u.spectral_density(wave[:, None, None]))
+
+            if factor == 1:    # Normalise to 1 arcsec2 if not a spatial density
+                # ..todo: lower needed because "DEG" is not understood, this is ugly
+                pixarea = (header['CDELT1'] * u.Unit(header['CUNIT1'].lower()) *
+                           header['CDELT2'] * u.Unit(header['CUNIT2'].lower())).to(u.arcsec**2)
+                data = data / pixarea.value    # cube is per arcsec2
+
+            data = (data * factor).value
+
+            cube.header['BUNIT'] = 'PHOTLAM/arcsec2'    # ..todo: make this more explicit?
+
+            # The imageplane_utils like to have the spatial WCS in units of "deg". Ensure
+            # that the cube is passed on accordingly
+            cube.header['CDELT1'] = header['CDELT1'] * u.Unit(header['CUNIT1'].lower()).to(u.deg)
+            cube.header['CDELT2'] = header['CDELT2'] * u.Unit(header['CUNIT2'].lower()).to(u.deg)
+            cube.header['CUNIT1'] = 'deg'
+            cube.header['CUNIT2'] = 'deg'
+
+            # Put on fov wavegrid
+            # ..todo: This assumes that we have only one fov. Generalise?
```

Comment:
Currently we assume that there is a single fov (`fovs[0]`). This will have to be generalised later on.

Reply:
Issue  #99 , not required for accepting this pull request.